File: Dynamic Programming/Coin Change/recursive.java
public class recursive {
    public int helper(int[] coins, int amount) {
        if (amount == 0) return 0;

        int res = (int) 1e9;
        for (int coin : coins) {
            if (amount - coin >= 0) {
                res = Math.min(res, 1 + helper(coins, amount - coin));
            }
        }
        return res;
    }

    public int coinChange(int[] coins, int amount) {
        int minCoins = helper(coins, amount);
        return (minCoins >= 1e9) ? -1 : minCoins;
    }
}

File: Dynamic Programming/Coin Change/recursive.cpp
class Solution {
    public:
        int solve(int amt, vector<int>& coins){
            if(amt == 0)
                return 0;
            
            int ans = INT_MAX;
            for(int coin : coins){
                if(amt-coin >= 0) //ensuring amt stays positive
                    ans = min(ans + 0LL, 1LL + solve(amt-coin, coins));
            }
            return ans;
        }
        int coinChange(vector<int>& coins, int amount) {
            int ans = solve(amount, coins);
    
            return ans==INT_MAX ? -1 : ans;
        }
    };